### PR TITLE
sync_diff_inspector: merge branch sync_diff_inspector 2.0 into master

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve? 

merge branch sync_diff_inspector 2.0 into master

### What is changed and how it works?

restructure sync_diff_inspector

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: tests/sync_diff_inspector/table_config/config.toml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    # 3 summary: summary.txt
+    # 4 checkpoint: a dir
+    output-dir = "/tmp/tidb_tools_test/sync_diff_inspector/output"
+
+    source-instances = ["tidb1"]
+
+    target-instance = "tidb"
+
+    # tables need to check.
+	target-check-tables = ["diff_test.test"]
```
Comment (3pointer) [inline tests/sync_diff_inspector/table_config/config.toml:64]:
```suggestion
    target-check-tables = ["diff_test.test"]
```

File: tests/sync_diff_inspector/table_config/config.toml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    # extra table config
+	target-configs= ["config1"]
```
Comment (3pointer) [inline tests/sync_diff_inspector/table_config/config.toml:67]:
```suggestion
    target-configs= ["config1"]
```

File: sync_diff_inspector/config/dm.go [hunk 3]
```diff
@@ -36,7 +36,7 @@ func getDMTaskCfgURL(dmAddr, task string) string {
 	return fmt.Sprintf("%s/apis/%s/subtasks/%s", dmAddr, apiVersion, task)
 }
 
-// getDMTaskCfg gets dm's sub task config
+// GetDMTaskCfg gets dm's sub task config
```
Comment (3pointer) [inline sync_diff_inspector/config/dm.go:39]:
```suggestion
// getDMTaskCfg gets dm's sub task config
```

File: tests/sync_diff_inspector/checkpoint/config_base_rand.toml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    # 3 summary: summary.txt
+    # 4 checkpoint: a dir
+    output-dir = "/tmp/tidb_tools_test/sync_diff_inspector/output"
+
+    source-instances = ["mysql1"]
+
+    target-instance = "mysql"
+
+    # tables need to check.
+	target-check-tables = ["diff_test.test"]
```
Comment (3pointer) [inline tests/sync_diff_inspector/checkpoint/config_base_rand.toml:62]:
```suggestion
    target-check-tables = ["diff_test.test"]
```

File: tests/sync_diff_inspector/checkpoint/config_base_rand.toml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    # extra table config
+	target-configs= ["config1"]
```
Comment (3pointer) [inline tests/sync_diff_inspector/checkpoint/config_base_rand.toml:65]:
```suggestion
    target-configs= ["config1"]
```

File: tests/sync_diff_inspector/config_base_mysql.toml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    # 3 summary: summary.txt
+    # 4 checkpoint: a dir
+    output-dir = "/tmp/tidb_tools_test/sync_diff_inspector/output"
+
+    source-instances = ["mysql1"]
+
+    target-instance = "tidb"
+
+    # tables need to check.
+	target-check-tables = ["diff_test.test"]
```
Comment (3pointer) [inline tests/sync_diff_inspector/config_base_mysql.toml:57]:
```suggestion
    target-check-tables = ["diff_test.test"]
```

File: tests/sync_diff_inspector/config_base_tidb.toml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    # 3 summary: summary.txt
+    # 4 checkpoint: a dir
+    output-dir = "/tmp/tidb_tools_test/sync_diff_inspector/output"
+
+    source-instances = ["tidb1"]
+
+    target-instance = "tidb"
+
+    # tables need to check.
+	target-check-tables = ["diff_test.test"]
```
Comment (3pointer) [inline tests/sync_diff_inspector/config_base_tidb.toml:57]:
```suggestion
    target-check-tables = ["diff_test.test"]
```

File: tests/sync_diff_inspector/shard/config_base.toml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    # 3 summary: summary.txt
+    # 4 checkpoint: a dir
+    output-dir = "/tmp/tidb_tools_test/sync_diff_inspector/output"
+
+    source-instances = ["mysql1"]
+
+    target-instance = "tidb"
+
+    # tables need to check.
+	target-check-tables = ["shard_test.test"]
```
Comment (3pointer) [inline tests/sync_diff_inspector/shard/config_base.toml:64]:
```suggestion
    target-check-tables = ["shard_test.test"]
```

File: tests/sync_diff_inspector/snapshot/config_base.toml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    # 3 summary: summary.txt
+    # 4 checkpoint: a dir
+    output-dir = "/tmp/tidb_tools_test/sync_diff_inspector/output"
+
+    source-instances = ["tidb1"]
+
+    target-instance = "tidb"
+
+    # tables need to check.
+	target-check-tables = ["diff_test.test"]
```
Comment (3pointer) [inline tests/sync_diff_inspector/snapshot/config_base.toml:57]:
```suggestion
    target-check-tables = ["diff_test.test"]
```

File: sync_diff_inspector/checkpoints/checkpoints.go [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+	"os"
+	"sync"
+
+	"github.com/pingcap/tidb-tools/sync_diff_inspector/config"
+	"github.com/pingcap/tidb-tools/sync_diff_inspector/report"
+
+	"github.com/pingcap/tidb-tools/sync_diff_inspector/chunk"
+	"github.com/siddontang/go/ioutil2"
+
+	//"github.com/golang/protobuf/proto"
```
Comment (3pointer) [inline sync_diff_inspector/checkpoints/checkpoints.go:29]:
```suggestion
```

File: sync_diff_inspector/checkpoints/checkpoints.go [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+	//"github.com/pingcap/errors"
```
Comment (3pointer) [inline sync_diff_inspector/checkpoints/checkpoints.go:30]:
```suggestion
```

File: sync_diff_inspector/config/config.toml [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+
+######################### Databases config #########################
+[data-sources]
+[data-sources.mysql1]
+    host = "127.0.0.1"
+    port = 3306
+    user = "root"
+    password = ""
+    # remove comment if use tidb's snapshot data
```
Comment (3pointer) [inline sync_diff_inspector/config/config.toml:33]: mysql doesn't has snapshot config

**Comment by Leavrth** [comment]
need to resolve conflict

**Pull request closed without merging**